File: sdk/swift/Sources/HiTechCloudSDK/CloudService.swift
import Foundation

/// Cloud Service: VM shutdown, reset, hostname, IP pools, networks, interfaces
public class CloudServiceResource {
    private let httpClient: HTTPClient
    public init(httpClient: HTTPClient) { self.httpClient = httpClient }

    /// POST /api/service/{id}/vms/{vmid}/shutdown - Perform graceful shutdown
    public func shutdownVm(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/vms/{vmid}/shutdown", params: params)
    }

    /// POST /api/service/{id}/vms/{vmid}/reset - Reset virtual server power
    public func resetVm(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/vms/{vmid}/reset", params: params)
    }

    /// POST /api/service/{id}/vms/{vmid}/hostname - Update VM hostname
    public func changeHostname(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/vms/{vmid}/hostname", params: params)
    }

    /// GET /api/service/{id}/vms/{vmid}/ippool - List IP pools
    public func listIpPools(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/vms/{vmid}/ippool", params: params)
    }

    /// POST /api/service/{id}/vms/{vmid}/ippool/{pool} - Allocate new IP
    public func allocateIp(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/vms/{vmid}/ippool/{pool}", params: params)
    }

    /// GET /api/service/{id}/networks - List available networks
    public func listNetworks(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/networks", params: params)
    }

    /// POST /api/service/{id}/vms/{vmid}/interfaces - Add network interface
    public func addInterface(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.post("/api/service/{id}/vms/{vmid}/interfaces", params: params)
    }

    /// GET /api/service/{id}/vms/{vmid}/interfaces/{iface} - Get interface details
    public func getInterface(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.get("/api/service/{id}/vms/{vmid}/interfaces/{iface}", params: params)
    }

    /// PUT /api/service/{id}/vms/{vmid}/interfaces/{iface} - Update interface
    public func updateInterface(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.put("/api/service/{id}/vms/{vmid}/interfaces/{iface}", params: params)
    }

    /// DELETE /api/service/{id}/vms/{vmid}/interfaces/{iface} - Remove interface
    public func removeInterface(params: [String: Any?] = [:]) throws -> [String: Any] {
        return try httpClient.delete("/api/service/{id}/vms/{vmid}/interfaces/{iface}", params: params)
    }

}

File: sdk/kotlin/src/main/kotlin/vn/hitechcloud/sdk/CloudService.kt
package vn.hitechcloud.sdk

/** Cloud Service: VM shutdown, reset, hostname, IP pools, networks, interfaces */
class CloudServiceResource(private val httpClient: HttpClient) {
    /** POST /api/service/{id}/vms/{vmid}/shutdown - Perform graceful shutdown */
    fun shutdownVm(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/vms/{vmid}/shutdown", params)
    }

    /** POST /api/service/{id}/vms/{vmid}/reset - Reset virtual server power */
    fun resetVm(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/vms/{vmid}/reset", params)
    }

    /** POST /api/service/{id}/vms/{vmid}/hostname - Update VM hostname */
    fun changeHostname(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/vms/{vmid}/hostname", params)
    }

    /** GET /api/service/{id}/vms/{vmid}/ippool - List IP pools */
    fun listIpPools(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/vms/{vmid}/ippool", params)
    }

    /** POST /api/service/{id}/vms/{vmid}/ippool/{pool} - Allocate new IP */
    fun allocateIp(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/vms/{vmid}/ippool/{pool}", params)
    }

    /** GET /api/service/{id}/networks - List available networks */
    fun listNetworks(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/networks", params)
    }

    /** POST /api/service/{id}/vms/{vmid}/interfaces - Add network interface */
    fun addInterface(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.post("/api/service/{id}/vms/{vmid}/interfaces", params)
    }

    /** GET /api/service/{id}/vms/{vmid}/interfaces/{iface} - Get interface details */
    fun getInterface(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.get("/api/service/{id}/vms/{vmid}/interfaces/{iface}", params)
    }

    /** PUT /api/service/{id}/vms/{vmid}/interfaces/{iface} - Update interface */
    fun updateInterface(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.put("/api/service/{id}/vms/{vmid}/interfaces/{iface}", params)
    }

    /** DELETE /api/service/{id}/vms/{vmid}/interfaces/{iface} - Remove interface */
    fun removeInterface(params: Map<String, Any?> = emptyMap()): Map<String, Any> {
        return httpClient.delete("/api/service/{id}/vms/{vmid}/interfaces/{iface}", params)
    }

}
